lib/middlewares/src/metering.rs

```diff
@@ -36,53 +41,63 @@ pub struct FunctionMetering<F: Fn(&Operator) -> u64 + Copy + Clone + Send + Sync
     /// Function that maps each operator to a cost in "points".
     cost_function: F,
 
+    /// The global index in the current module for the error code.
+    error_code_index: GlobalIndex,
+
     /// The global index in the current module for remaining points.
     remaining_points_index: GlobalIndex,
 
     /// Accumulated cost of the current basic block.
     accumulated_cost: u64,
 }
 
+#[derive(Debug, PartialEq)]
+pub enum MeteringError {
+    NoError,
+    OutOfGas,
+}
+
+impl TryFrom<i32> for MeteringError {
+    type Error = ();
+
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        match value {
+            value if value == MeteringError::NoError as _ => Ok(MeteringError::NoError),
+            value if value == MeteringError::OutOfGas as _ => Ok(MeteringError::OutOfGas),
+            _ => Err(()),
+        }
+    }
+}
+
+impl<T> TryFrom<Value<T>> for MeteringError {
+    type Error = ();
+
+    fn try_from(v: Value<T>) -> Result<Self, Self::Error> {
+        match v {
+            Value::I32(value) => value.try_into(),
+            _ => Err(()),
+        }
```

Comment:
We could also convert to `u32` because `impl<T> TryFrom<Value<T>> for u32` already exists. Then we cam make the `MeteringError` enum `#[repr(u32)]` and avoid a lot of casting